src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,400 @@
+        }
+        this.hmacAlgName = hmacAlgName;
+        this.hmacLen = hmacLen;
+    }
+
+    /**
+     * Derive a key, returned as a {@code SecretKey}.
+     *
+     * @return a derived {@code SecretKey} object of the specified algorithm
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationParameterSpec} is
+     *     invalid or if the combination of {@code alg} and the {@code derivationParameterSpec}
+     *     results in something invalid, ie - a key of inappropriate length
+     *     for the specified algorithm
+     * @throws NoSuchAlgorithmException
+     *     if {@code alg} is empty or invalid
+     * @throws IllegalArgumentException
+     *     if {@code alg} is {@code null} or empty
+     */
+    @Override
+    protected SecretKey engineDeriveKey(String alg,
+                                        AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException, NoSuchAlgorithmException {
+
+        if (alg == null) {
+            throw new NullPointerException(
+                "the algorithm for the SecretKey return value must not be null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                "the algorithm for the SecretKey return value must not be "
+                + "empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(derivationParameterSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived keying material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(AlgorithmParameterSpec derivationParameterSpec)
+        throws InvalidAlgorithmParameterException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        SecretKey inputKeyMaterial;
+        SecretKey salt;
+        SecretKey pseudoRandomKey;
+        byte[] info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (derivationParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) derivationParameterSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine these Lists of key segments into
+            // single SecretKey Objects, unless we were passed something bogus
+            // or an unexportable P11 key
```

Comment:
Still has typo "Objects", should be "object".
BTW, I still strongly feel that it can be combined into a byte[] instead of a SecretKey object.